### PR TITLE
Performance improvements for collect_2q_blocks

### PR DESCRIPTION
**Pull request by IvanIsCoding**
### Summary

Closes https://github.com/Qiskit/retworkx/issues/265

This PR contains two core improvements that boost the performance of collect_2q_blocks:
* Use new functions from retworkx in `quantum_successors` and `quantum_predecessors` that are more efficient
* Add `is_successor` and `is_predecessor` function to `DAGCircuit` to efficiently check for successors/predecessors instead of getting all successors/predecessors and then doing a check

It also contains two minor performance optimizations that are open to debate:
* Replace `qargs` with `_qargs`, because of `@property` overhead when getting `qargs`
* Replace `op` with `_op`, idem

### Details and comments

This PR also  upgrades retworkx to version 0.9

### CODE REVIEW

**Comment by IvanIsCoding** [comment]
I will edit this comment later with more details about benchmarks. It gives a nice boost without a full Rust rewrite, with a 1.3x speed improvement on my local machine.

File: requirements.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 contextvars>=2.4;python_version<'3.7'
 jsonschema>=2.6
-retworkx>=0.8.0
+retworkx @ git+https://github.com/Qiskit/retworkx@main
```
Comment (mtreinish) [inline requirements.txt:3]: retworkx 0.9.0 has been released now so we can update this:

```suggestion
retworkx>=0.9.0
```

**Review by mtreinish — COMMENTED** [review]
Can you add an `upgrade` release note here about the bump in the minimum required retworkx version. We should call that out in the release notes as a thing that might require end user manual updates when upgrading to the next terra release.

**Comment by IvanIsCoding** [comment]
> Can you add an `upgrade` release note here about the bump in the minimum required retworkx version. We should call that out in the release notes as a thing that might require end user manual updates when upgrading to the next terra release.

I added the release note. This is no longer WIP, it can be merged

File: releasenotes/notes/bump-retworkx-59c7ab59ad02c3c0.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+upgrade:
+  - |
+    `retworkx <https://pypi.org/project/retworkx/>`_ dependency was upgraded to version `0.9`
```
Comment (mtreinish) [inline releasenotes/notes/bump-retworkx-59c7ab59ad02c3c0.yaml:3]:
```suggestion
features:
  - |
    Added two new methods, :meth:`~qiskit.dagcircuit.DAGCircuit.is_successor` and
    :meth:`~qiskit.dagcircuit.DAGCircuit.is_predecessor`, to the
    :class:`~qiskit.dagcircuit.DAGCircuit` class. These functions are used to check if a node
    is either a successor or predecessor of another node on the
     :class:`~qiskit.dagcircuit.DAGCircuit`.
upgrade:
  - |
    The minimum version of the `retworkx <https://pypi.org/project/retworkx/>`_ dependency
    was increased to version `0.9.0`. This was done to use new APIs introduced in that release
    which improved the performance of some transpiler passes.
```

**Review by mtreinish — COMMENTED** [review]
LGTM, just one comment with some suggestions on the release notes. Other than that I think this is ready to go.

**Comment by mtreinish** [comment]
I also ran the asv collect_2q_blocks benchmark on this and it yielded this result:

```
Benchmarks that have improved:

       before           after         ratio
     [a2193a5c]       [e2935d73]
     <main>       <IvanIsCoding-improve_collect2q>
-      44.3±0.2ms       30.4±0.2ms     0.69  passes.PassBenchmarks.time_collect_2q_blocks(14, 1024)
-      65.8±0.3ms       44.9±0.6ms     0.68  passes.PassBenchmarks.time_collect_2q_blocks(20, 1024)
-      15.4±0.1ms       10.3±0.3ms     0.67  passes.PassBenchmarks.time_collect_2q_blocks(5, 1024)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by mtreinish** [comment]
I also ran a quick benchmark on https://github.com/Qiskit/qiskit/blob/master/test/benchmarks/qasm/test_eoh_qasm.qasm (which is different load on the pass because it's all 2q, so one giant block) with:

```python
import time
import statistics

from qiskit.transpiler.passes import Collect2qBlocks
from qiskit.circuit import QuantumCircuit
from qiskit.converters import circuit_to_dag

qc = QuantumCircuit.from_qasm_file('test_eoh_qasm.qasm')
dag = circuit_to_dag(qc)
collect_pass = Collect2qBlocks()
times = []
for i in range(20):
    start = time.time()
    collect_pass.run(dag)
    stop = time.time()
    times.append(stop - start)
print("Mean run time: %s" % statistics.mean(times))
```

which returns 0.015 seconds on `main` and 0.01 seconds with this PR.

**Comment by IvanIsCoding** [comment]
I commited the suggestions to the release note, the need for retworkx upgrade should be clearer now

> I also ran the asv c[ollect_2q_blocks benchmark](https://github.com/Qiskit/qiskit/blob/master/test/benchmarks/passes.py#L175-L176) on this and it yielded this result:
> 
> ```
> Benchmarks that have improved:
> 
>        before           after         ratio
>      [a2193a5c]       [e2935d73]
>      <main>       <IvanIsCoding-improve_collect2q>
> -      44.3±0.2ms       30.4±0.2ms     0.69  passes.PassBenchmarks.time_collect_2q_blocks(14, 1024)
> -      65.8±0.3ms       44.9±0.6ms     0.68  passes.PassBenchmarks.time_collect_2q_blocks(20, 1024)
> -      15.4±0.1ms       10.3±0.3ms     0.67  passes.PassBenchmarks.time_collect_2q_blocks(5, 1024)
> 
> SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
> PERFORMANCE INCREASED.
> ```

Cool, the 1.3x - 1.5x speed up remained consistent accross the benchmarks.

**Review by mtreinish — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]
Oops, thanks for fixing the whitespace issue in my release note suggestion (sigh getting whitespace right in the github editor is so irritating).

**Pull request merged**